lib/serialization/compositeSpec.ts

```diff
@@ -8,29 +8,94 @@ export interface CompositeType {
   [key: string]: any;
 }
 
+export interface PolymorphicDerivedType {
+  /**
+   * A CompositeTypeSpec that "inherits" from this CompositeTypeSpec.
+   */
+  derivedTypeSpec: CompositeTypeSpec | string;
+
+  /**
+   * The value of the discriminator property that would indicate that this CompositeTypeSpec is the
+   * target TypeSpec for the value being serialized or deserialized.
+   */
+  discriminatorPropertyValue: any;
+}
+
+/**
+ * The properties that specify the derived types of a CompositeSpecType.
+ */
+export interface PolymorphicDerivedTypes {
+  /**
+   * The CompositeTypeSpecs that "inherit" from this CompositeTypeSpec.
+   */
+  derivedTypes: PolymorphicDerivedType[];
+
+  /**
+   * The name of the property that determines what the actual type the polymorphic object is.
+   */
+  discriminatorPropertyName: string;
+
+  /**
+   * The serialized name of the property that determines what the actual type of the polymorphic
+   * object is. If this is not provided, then the discriminatorPropertyName should be used.
+   */
+  discriminatorPropertySerializedName?: string;
+
+  /**
+   * The value of the discriminator property that would indicate that this CompositeTypeSpec is the
+   * target TypeSpec for the value being serialized or deserialized.
+   */
+  discriminatorPropertyValue: any;
+}
+
+/**
+ * The options that specify polymorphism for a CompositeTypeSpec.
+ */
+export interface Polymorphism {
+  /**
+   * The base types for this CompositeTypeSpec.
+   */
+  inheritsFrom?: (CompositeTypeSpec | string) | (CompositeTypeSpec | string)[];
```

Comment:
The parentheses make it a little confusing while viewing. In reality it could just be oneOf the 4 combinations. So how about doing something like this:
```js
CompositeTypeSpec | string | CompositeTypeSpec[] | string[]
```

Reply:
I think `(CompositeTypeSpec | string)[]` is a different type than `CompositeTypeSpec[] | string[]`. But I might not fully understand how `|` distributes in TypeScript.

Reply:
I actually think removing the first set of parens would help readability.

Reply:
I'll remove the first set of parentheses. @RikkiGibson is correct that the intent is that it can be an array of a combination of CompositeTypeSpecs and strings.

Reply:
Fixed in https://github.com/Azure/ms-rest-js/pull/63/commits/f40e92e22aef9ee866814a758f67b0d7773a5c3f.

Reply:
how about doing it like this:
```
export type ParentTypeOrParentName = CompositeType | string;

inheritsFrom?: ParentTypeOrParentName | ParentTypeOrParentName[]
```

Reply:
Done in https://github.com/Azure/ms-rest-js/pull/63/commits/f65d798c55cb4262eb43feb80afbba9a39a6c5fc.